clippy_lints/src/methods/manual_inspect.rs

```diff
@@ -207,36 +211,75 @@ enum UseKind<'tcx> {
     {
-        return (UseKind::AutoBorrowed, use_cx.same_ctxt);
+        return (UseKind::AutoBorrowed, mutbl, use_cx.same_ctxt);
     }
     let res = match use_cx.use_node(cx) {
         ExprUseNode::Return(_) => {
             if let ExprKind::Ret(Some(e)) = use_cx.node.expect_expr().kind {
                 UseKind::Return(e.span)
             } else {
-                return (UseKind::Return(DUMMY_SP), false);
+                return (UseKind::Return(DUMMY_SP), mutbl, false);
             }
         },
-        ExprUseNode::FieldAccess(name) => UseKind::FieldAccess(name.name, use_cx.node.expect_expr()),
+        ExprUseNode::FieldAccess(_, name) => UseKind::FieldAccess(name.name, use_cx.node.expect_expr()),
         ExprUseNode::Callee | ExprUseNode::MethodArg(_, _, 0)
             if use_cx
                 .adjustments
                 .first()
-                .is_some_and(|a| matches!(a.kind, Adjust::Borrow(AutoBorrow::Ref(AutoBorrowMutability::Not)))) =>
+                .is_some_and(|a| matches!(a.kind, Adjust::Borrow(AutoBorrow::Ref(_)))) =>
         {
             UseKind::AutoBorrowed
         },
         ExprUseNode::Callee | ExprUseNode::MethodArg(_, _, 0) => UseKind::WillAutoDeref,
         ExprUseNode::AddrOf(BorrowKind::Ref, _) => UseKind::Borrowed(use_cx.node.expect_expr().span),
         _ => UseKind::Deref,
     };
-    (res, use_cx.same_ctxt)
+    (res, mutbl, use_cx.same_ctxt)
+}
+
+fn use_mutability(cx: &LateContext<'_>, expr_id: HirId) -> Mutability {
+    let adjusted = |expr: &Expr<'_>| -> Mutability {
+        let adj = cx.typeck_results().expr_adjustments(expr);
+        if let Some(Adjust::Borrow(AutoBorrow::Ref(mutbl))) = adj.last().map(|adj| &adj.kind) {
+            (*mutbl).into()
+        } else {
+            Mutability::Not
+        }
+    };
+
+    let mut last_child = None;
+
+    for (_, node) in cx.tcx.hir().parent_iter(expr_id) {
+        if let Node::Expr(expr) = node {
+            match expr.kind {
+                ExprKind::AddrOf(_, mutbl, _) => return mutbl,
+                ExprKind::MethodCall(_, self_arg, _, _) => return adjusted(self_arg),
+                ExprKind::Call(f, args) => return adjusted(args.iter().find(|arg| arg.hir_id == expr_id).unwrap_or(f)),
+                ExprKind::Field(field, _) | ExprKind::Index(field, _, _) => last_child = Some(field.hir_id),
+                ExprKind::Assign(lhs, _, _) | ExprKind::AssignOp(_, lhs, _) => {
+                    let is_lhs = match lhs.kind {
+                        ExprKind::Field(child, _) | ExprKind::Index(child, _, _) => {
+                            last_child.is_some_and(|field_id| field_id == child.hir_id)
+                        },
+                        _ => false,
+                    };
+                    if is_lhs {
+                        return Mutability::Mut;
+                    }
+                    return Mutability::Not;
+                },
+                _ => {},
+            }
+        }
+    }
```

Comment:
This loop has a few errors.
* The loop should stop at the first unmatched node and adjustments should be checked here.
* The current child id should be initialized to the starting id and updated before looping so it can be used for checking adjustments.
* Checking the lhs for assignment is then just `child_id == lhs.hir_id`.
* `MethodCall` and `Call` don't need to be special cased.